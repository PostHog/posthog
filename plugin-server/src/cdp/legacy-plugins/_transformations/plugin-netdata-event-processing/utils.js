export function cleanPropertyName(k) {
    return k
        // convert to lower case
        .toLowerCase()
        // remove leading slash
        .replace(/^\//, "")
        // replace all slashes and dots with _
        .replace(/\/|\.|-| /g, "_")
        ;
}

export function isStringDDMMYYYYHHMM(dt){
    var reDate = /^((0?[1-9]|[12][0-9]|3[01])[- /.](0?[1-9]|1[012])[- /.](19|20)?[0-9]{2}[ ][012][0-9][:][0-9]{2})*$/;
    return reDate.test(dt);
}

export function isDemo(url) {
    if (
        (url.includes('://london.my-netdata.io'))
        ||
        (url.includes('://london3.my-netdata.io'))
        ||
        (url.includes('://cdn77.my-netdata.io'))
        ||
        (url.includes('://octopuscs.my-netdata.io'))
        ||
        (url.includes('://bangalore.my-netdata.io'))
        ||
        (url.includes('://frankfurt.my-netdata.io'))
        ||
        (url.includes('://newyork.my-netdata.io'))
        ||
        (url.includes('://sanfrancisco.my-netdata.io'))
        ||
        (url.includes('://singapore.my-netdata.io'))
        ||
        (url.includes('://toronto.my-netdata.io'))
        ){
        return true
    } else {
        return false
    }
}

export function splitPathName(event) {
    if (event.properties['$pathname']) {
        event.properties["$pathname"].split("/").forEach((pathname, index) => {
          if ((pathname !== "") && (pathname !== null)){
            event.properties[`pathname_${index}`] = pathname
          }
        })
      }
    return event
}