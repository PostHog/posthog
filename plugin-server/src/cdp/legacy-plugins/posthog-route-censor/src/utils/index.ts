export * from './checkIsValidHttpUrl'
export * from './censorUrlPath'
