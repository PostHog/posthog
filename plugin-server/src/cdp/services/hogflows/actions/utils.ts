import { HogFlow, HogFlowAction } from '~/schema/hogflow'

export const findActionById = (hogFlow: HogFlow, id: string): HogFlowAction => {
    const action = hogFlow.actions.find((action) => action.id === id)
    if (!action) {
        throw new Error(`Action ${id} not found`)
    }

    return action
}

export const findActionByType = <T extends HogFlowAction['type']>(
    hogFlow: HogFlow,
    type: T
): Extract<HogFlowAction, { type: T }> | undefined => {
    const action = hogFlow.actions.find((action) => action.type === type)
    if (!action) {
        return undefined
    }

    return action as Extract<HogFlowAction, { type: T }>
}

export const findNextAction = (
    hogFlow: HogFlow,
    currentActionId: string,
    edgeIndex?: number
): HogFlowAction | undefined => {
    const edges = hogFlow.edges.filter((edge) => edge.from === currentActionId)

    let nextActionId: string | undefined

    if (edgeIndex === undefined) {
        nextActionId = edges.find((edge) => edge.type === 'continue')?.to
    } else {
        nextActionId = edges.find((edge) => edge.type === 'branch' && edge.index === edgeIndex)?.to
    }

    return nextActionId ? findActionById(hogFlow, nextActionId) : undefined
}
