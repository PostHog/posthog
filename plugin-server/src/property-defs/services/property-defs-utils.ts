import { DateTime } from 'luxon'

import { PropertyType } from '../../types'

// lifted from here:
// https://github.com/PostHog/posthog/blob/021aaab04b4acd96cf8121c033ac3b0042492598/rust/property-defs-rs/src/types.rs#L457-L461
const DJANGO_MAX_CHARFIELD_LENGTH = 200

// These properties have special meaning, and are ignored
export const PROPERTY_DEFS_PROPERTIES_TO_SKIP: string[] = [
    '$set',
    '$set_once',
    '$unset',
    '$group_0',
    '$group_1',
    '$group_2',
    '$group_3',
    '$group_4',
    '$groups',
]

export const PROPERTY_DEFS_DATE_PROP_KEYWORDS: string[] = [
    'time',
    'timestamp',
    'date',
    '_at',
    '-at',
    'createdat',
    'updatedat',
]

export function willFitInPostgres(s: string) {
    return s.length < DJANGO_MAX_CHARFIELD_LENGTH
}

export function sanitizeEventName(eventName: string) {
    return eventName.replace('\u0000', '\uFFFD')
}

export function sixMonthsAgoUnixSeconds() {
    const now = new Date()
    now.setMonth(now.getMonth() - 6)
    return Math.floor(now.getTime() / 1000)
}

const BOOLEAN_VALUES = ['true', 'false', 'TRUE', 'FALSE']

export const getPropertyType = (rawKey: string, value: any): PropertyType | null => {
    const key = rawKey.trim().toLowerCase()

    // Special cases for certain property prefixes
    if (key.startsWith('utm_')) {
        // utm_ prefixed properties should always be detected as strings.
        // Sometimes the first value sent looks like a number, even though
        // subsequent values are not.
        return PropertyType.String
    }
    if (key.startsWith('$feature/')) {
        // $feature/ prefixed properties should always be detected as strings.
        // These are feature flag values, and can be boolean or string.
        // Sometimes the first value sent is boolean (because flag isn't enabled) while
        // subsequent values are not. We don't want this to be misunderstood as a boolean.
        return PropertyType.String
    }

    if (key === '$feature_flag_response') {
        // $feature_flag_response properties should always be detected as strings.
        // These are feature flag values, and can be boolean or string.
        // Sometimes the first value sent is boolean (because flag isn't enabled) while
        // subsequent values are not. We don't want this to be misunderstood as a boolean.
        return PropertyType.String
    }

    if (key.startsWith('$survey_response')) {
        // NB: $survey_responses are collected in an interesting way, where the first
        // response is called `$survey_response` and subsequent responses are called
        // `$survey_response_2`, `$survey_response_3`, etc. So, this check should auto-cast
        // all survey responses to strings.
        return PropertyType.String
    }

    if (typeof value === 'string') {
        const s = value.trim()
        if (BOOLEAN_VALUES.includes(s)) {
            return PropertyType.Boolean
        }
        // Try to parse this as an ISO 8601 date
        try {
            if (PROPERTY_DEFS_DATE_PROP_KEYWORDS.some((kw) => key.includes(kw))) {
                return PropertyType.DateTime
            }
            const date = DateTime.fromISO(s)
            if (date.isValid) {
                return PropertyType.DateTime
            }
            // TODO(eli): add speculative date string matching?
        } catch {
            // Not a valid date, continue to string type
        }
        return PropertyType.String
    }

    if (typeof value === 'boolean') {
        return PropertyType.Boolean
    }

    if (typeof value === 'number') {
        if (value >= sixMonthsAgoUnixSeconds()) {
            return PropertyType.DateTime
        }
        return PropertyType.Numeric
    }

    return null
}
