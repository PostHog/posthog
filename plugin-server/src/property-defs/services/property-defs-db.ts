import { DateTime } from 'luxon'

import { EventDefinitionType, EventPropertyType, Hub, PropertyDefinitionType } from '../../types'
import { PostgresUse } from '../../utils/db/postgres'
import { status } from '../../utils/status'

export class PropertyDefsDB {
    constructor(private hub: Hub) {}

    async writeEventProperty(eventProperty: EventPropertyType) {
        await this.hub.postgres
            .query(
                PostgresUse.COMMON_WRITE,
                `INSERT INTO posthog_eventproperty (event, property, team_id, project_id)
                    VALUES ($1, $2, $3, $4)
                    ON CONFLICT DO NOTHING
            `,
                [eventProperty.event, eventProperty.property, eventProperty.team_id, eventProperty.project_id],
                'upsertEventProperty'
            )
            .catch((e) => {
                status.error('游대', `Error writing event property`, { eventProperty, error: e.message })
                throw e
            })
    }

    async writeEventProperties(eventProperties: EventPropertyType[]) {
        await this.hub.postgres
            .query(
                PostgresUse.COMMON_WRITE,
                `INSERT INTO posthog_eventproperty (event, property, team_id, project_id)
                    VALUES (UNNEST($1::text[]), UNNEST($2::text[]), UNNEST($3::int[]), UNNEST($4::int[]))
                    ON CONFLICT DO NOTHING
                `,
                [
                    eventProperties.map((ep) => ep.event),
                    eventProperties.map((ep) => ep.property),
                    eventProperties.map((ep) => ep.team_id),
                    eventProperties.map((ep) => ep.project_id),
                ],
                'upsertEventPropertiesBatch'
            )
            .catch((e) => {
                status.error('游대', `Error writing event properties batch`, { eventProperties, error: e.message })
                throw e
            })
    }

    async writePropertyDefinition(propertyDefinition: PropertyDefinitionType) {
        await this.hub.postgres
            .query(
                PostgresUse.COMMON_WRITE,
                `INSERT INTO posthog_propertydefinition (id, name, type, group_type_index, is_numerical, team_id, project_id, property_type, volume_30_day, query_usage_30_day)
                    VALUES ($1, $2, $3, $4, $5, $6, $7, $8, NULL, NULL)
                    ON CONFLICT (coalesce(project_id, team_id::bigint), name, type, coalesce(group_type_index, -1))
                    DO UPDATE SET property_type=EXCLUDED.property_type
                    WHERE posthog_propertydefinition.property_type IS NULL`,
                [
                    propertyDefinition.id,
                    propertyDefinition.name,
                    propertyDefinition.type,
                    propertyDefinition.group_type_index,
                    propertyDefinition.is_numerical,
                    propertyDefinition.team_id,
                    propertyDefinition.project_id,
                    propertyDefinition.property_type,
                ],
                'upsertPropertyDefinition'
            )
            .catch((e) => {
                status.error('游대', `Error writing property definitions batch`, {
                    propertyDefinition,
                    error: e.message,
                })
                throw e
            })
    }

    async writePropertyDefinitions(propertyDefinitions: PropertyDefinitionType[]) {
        await this.hub.postgres
            .query(
                PostgresUse.COMMON_WRITE,
                `INSERT INTO posthog_propertydefinition (id, name, type, group_type_index, is_numerical, team_id, project_id, property_type, volume_30_day, query_usage_30_day)
                    VALUES (UNNEST($1::uuid[]), UNNEST($2::text[]), UNNEST($3::smallint[]), UNNEST($4::int[]), UNNEST($5::boolean[]), UNNEST($6::int[]), UNNEST($7::int[]), UNNEST($8::text[]), NULL, NULL)
                    ON CONFLICT (coalesce(project_id, team_id::bigint), name, type, coalesce(group_type_index, -1))
                    DO UPDATE SET property_type=EXCLUDED.property_type
                    WHERE posthog_propertydefinition.property_type IS NULL
                `,
                [
                    propertyDefinitions.map((pd) => pd.id),
                    propertyDefinitions.map((pd) => pd.name),
                    propertyDefinitions.map((pd) => pd.type),
                    propertyDefinitions.map((pd) => pd.group_type_index),
                    propertyDefinitions.map((pd) => pd.is_numerical),
                    propertyDefinitions.map((pd) => pd.team_id),
                    propertyDefinitions.map((pd) => pd.project_id),
                    propertyDefinitions.map((pd) => pd.property_type),
                ],
                'upsertPropertyDefinitionsBatch'
            )
            .catch((e) => {
                status.error('游대', `Error writing property definitions batch`, {
                    propertyDefinitions,
                    error: e.message,
                })
                throw e
            })
    }

    async writeEventDefinition(eventDefinition: EventDefinitionType) {
        await this.hub.postgres
            .query(
                PostgresUse.COMMON_WRITE,
                `INSERT INTO posthog_eventdefinition (id, name, team_id, project_id, last_seen_at, created_at, volume_30_day, query_usage_30_day)
                    VALUES ($1, $2, $3, $4, $5, $6, NULL, NULL)
                    ON CONFLICT (coalesce(project_id, team_id::bigint), name)
                    DO UPDATE SET last_seen_at=EXCLUDED.last_seen_at WHERE posthog_eventdefinition.last_seen_at < EXCLUDED.last_seen_at
                `,
                [
                    eventDefinition.id,
                    eventDefinition.name,
                    eventDefinition.team_id,
                    eventDefinition.project_id,
                    DateTime.now().toISO(), // TODO: Should this be the event timestamp?
                    DateTime.now().toISO(),
                ],
                'upsertEventDefinition'
            )
            .catch((e) => {
                status.error('游대', `Error writing event definition`, { eventDefinition, error: e.message })
                throw e
            })
    }

    async writeEventDefinitions(eventDefinitions: EventDefinitionType[]) {
        const now = DateTime.now().toISO()
        await this.hub.postgres
            .query(
                PostgresUse.COMMON_WRITE,
                `INSERT INTO posthog_eventdefinition (id, name, team_id, project_id, last_seen_at, created_at, volume_30_day, query_usage_30_day)
                    VALUES (UNNEST($1::uuid[]), UNNEST($2::text[]), UNNEST($3::int[]), UNNEST($4::int[]), UNNEST($5::timestamp[]), UNNEST($6::timestamp[]), NULL, NULL)
                    ON CONFLICT (coalesce(project_id, team_id::bigint), name)
                    DO UPDATE SET last_seen_at=EXCLUDED.last_seen_at WHERE posthog_eventdefinition.last_seen_at < EXCLUDED.last_seen_at
                `,
                [
                    eventDefinitions.map((ed) => ed.id),
                    eventDefinitions.map((ed) => ed.name),
                    eventDefinitions.map((ed) => ed.team_id),
                    eventDefinitions.map((ed) => ed.project_id),
                    eventDefinitions.map(() => now),
                    eventDefinitions.map(() => now),
                ],
                'upsertEventDefinitionsBatch'
            )
            .catch((e) => {
                status.error('游대', `Error writing event definitions batch`, { eventDefinitions, error: e.message })
                throw e
            })
    }

    async listPropertyDefinitions(teamId: number): Promise<PropertyDefinitionType[]> {
        const result = await this.hub.postgres.query<PropertyDefinitionType>(
            PostgresUse.COMMON_READ,
            `SELECT * FROM posthog_propertydefinition WHERE team_id = $1`,
            [teamId],
            'listPropertyDefinitions'
        )

        return result.rows
    }

    async listEventDefinitions(teamId: number): Promise<EventDefinitionType[]> {
        const result = await this.hub.postgres.query<EventDefinitionType>(
            PostgresUse.COMMON_READ,
            `SELECT * FROM posthog_eventdefinition WHERE team_id = $1`,
            [teamId],
            'listEventDefinitions'
        )

        return result.rows
    }

    async listEventProperties(teamId: number): Promise<EventPropertyType[]> {
        const result = await this.hub.postgres.query<EventPropertyType>(
            PostgresUse.COMMON_READ,
            `SELECT * FROM posthog_eventproperty WHERE team_id = $1`,
            [teamId],
            'listEventProperties'
        )

        return result.rows
    }
}
