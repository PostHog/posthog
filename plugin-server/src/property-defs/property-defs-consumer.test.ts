import { DateTime } from 'luxon'
import { Message } from 'node-rdkafka'

import { mockProducer } from '~/tests/helpers/mocks/producer.mock'
import { forSnapshot as _forSnapshot } from '~/tests/helpers/snapshots'
import { getFirstTeam, resetTestDatabase } from '~/tests/helpers/sql'

import { insertHogFunction as _insertHogFunction } from '../cdp/_tests/fixtures'
import {
    ClickHouseEvent,
    Hub,
    ProjectId,
    PropertyDefinitionType,
    RawClickHouseEvent,
    Team,
    TeamId,
    TimestampFormat,
} from '../types'
import { closeHub, createHub } from '../utils/db/hub'
import { PostgresUse } from '../utils/db/postgres'
import { castTimestampOrNow } from '../utils/utils'
import { PropertyDefsConsumer } from './property-defs-consumer'
import { PropertyDefsDB } from './services/property-defs-db'

const DEFAULT_TEST_TIMEOUT = 5000
jest.setTimeout(DEFAULT_TEST_TIMEOUT)

const mockConsumer = {
    on: jest.fn(),
    commitSync: jest.fn(),
    commit: jest.fn(),
    queryWatermarkOffsets: jest.fn(),
    committed: jest.fn(),
    assignments: jest.fn(),
    isConnected: jest.fn(() => true),
    getMetadata: jest.fn(),
}

jest.mock('../../src/kafka/batch-consumer', () => {
    return {
        startBatchConsumer: jest.fn(() =>
            Promise.resolve({
                join: () => ({
                    finally: jest.fn(),
                }),
                stop: jest.fn(),
                consumer: mockConsumer,
            })
        ),
    }
})

let offsetIncrementer = 0

const createRawClickHouseEvent = (event: ClickHouseEvent): RawClickHouseEvent => {
    // Inverts the parsing for simpler tests
    return {
        ...event,
        properties: JSON.stringify(event.properties),
        timestamp: castTimestampOrNow(event.timestamp ?? null, TimestampFormat.ClickHouse),
        created_at: castTimestampOrNow(event.created_at ?? null, TimestampFormat.ClickHouse),
        elements_chain: JSON.stringify(event.elements_chain),
        person_created_at: castTimestampOrNow(event.person_created_at ?? null, TimestampFormat.ClickHouse),
        person_properties: JSON.stringify(event.person_properties),
        group0_created_at: castTimestampOrNow(event.group0_created_at ?? null, TimestampFormat.ClickHouse),
        group0_properties: event.group0_properties ? JSON.stringify(event.group0_properties) : undefined,
        group1_created_at: castTimestampOrNow(event.group1_created_at ?? null, TimestampFormat.ClickHouse),
        group1_properties: event.group1_properties ? JSON.stringify(event.group1_properties) : undefined,
        group2_created_at: castTimestampOrNow(event.group2_created_at ?? null, TimestampFormat.ClickHouse),
        group2_properties: event.group2_properties ? JSON.stringify(event.group2_properties) : undefined,
        group3_created_at: castTimestampOrNow(event.group3_created_at ?? null, TimestampFormat.ClickHouse),
        group3_properties: event.group3_properties ? JSON.stringify(event.group3_properties) : undefined,
        group4_created_at: castTimestampOrNow(event.group4_created_at ?? null, TimestampFormat.ClickHouse),
        group4_properties: event.group4_properties ? JSON.stringify(event.group4_properties) : undefined,
    }
}

const createClickHouseEvent = (event: Partial<ClickHouseEvent> = {}): ClickHouseEvent => {
    return {
        uuid: event.uuid ?? '123',
        event: event.event ?? '$pageview',
        team_id: event.team_id ?? 1,
        project_id: event.project_id ?? (1 as ProjectId),
        distinct_id: event.distinct_id ?? 'distinct_id_1',
        /** Person UUID. */
        person_id: event.person_id ?? undefined,

        timestamp: DateTime.now(),
        created_at: DateTime.now(),
        properties: event.properties ?? {},
        elements_chain: event.elements_chain ?? null,
        person_created_at: event.person_created_at ?? null,
        person_properties: event.person_properties ?? {},
        group0_properties: event.group0_properties ?? {},
        group1_properties: event.group1_properties ?? {},
        group2_properties: event.group2_properties ?? {},
        group3_properties: event.group3_properties ?? {},
        group4_properties: event.group4_properties ?? {},
        group0_created_at: event.group0_created_at ?? null,
        group1_created_at: event.group1_created_at ?? null,
        group2_created_at: event.group2_created_at ?? null,
        group3_created_at: event.group3_created_at ?? null,
        group4_created_at: event.group4_created_at ?? null,
        person_mode: event.person_mode ?? 'full',
    }
}

const createKafkaMessages: (events: ClickHouseEvent[]) => Message[] = (events) => {
    return events.map((event) => {
        // TRICKY: This is the slightly different format that capture sends
        return {
            value: Buffer.from(JSON.stringify(createRawClickHouseEvent(event))),
            size: 1,
            topic: 'test',
            offset: offsetIncrementer++,
            timestamp: DateTime.now().toMillis(),
            partition: 1,
        }
    })
}

describe('PropertyDefsConsumer', () => {
    let ingester: PropertyDefsConsumer
    let hub: Hub
    let fixedTime: DateTime
    let team: Team
    let propertyDefsDB: PropertyDefsDB

    const advanceTime = (amount: number) => {
        fixedTime = fixedTime.plus({ seconds: amount })
        jest.spyOn(Date, 'now').mockReturnValue(fixedTime.toMillis())
    }

    beforeEach(async () => {
        fixedTime = DateTime.fromObject({ year: 2025, month: 1, day: 1 }, { zone: 'UTC' })
        jest.spyOn(Date, 'now').mockReturnValue(fixedTime.toMillis())

        offsetIncrementer = 0
        await resetTestDatabase()
        hub = await createHub()
        team = await getFirstTeam(hub)

        // Create some default group type mappings
        const groupTypeMappings = [
            { group_type: 'group-a', group_type_index: 0 },
            { group_type: 'group-b', group_type_index: 1 },
        ]
        for (const mapping of groupTypeMappings) {
            await hub.postgres.query(
                PostgresUse.COMMON_WRITE,
                `INSERT INTO posthog_grouptypemapping (team_id, project_id, group_type, group_type_index) VALUES ($1, $2, $3, $4)`,
                [team.id, team.project_id, mapping.group_type, mapping.group_type_index],
                'createGroupTypeMappings'
            )
        }

        hub.PROPERTY_DEFS_CONSUMER_ENABLED_TEAMS = '*'
        ingester = new PropertyDefsConsumer(hub)

        hub.kafkaProducer = mockProducer
        propertyDefsDB = ingester['propertyDefsDB']

        jest.spyOn(propertyDefsDB, 'writeEventDefinitions')
        jest.spyOn(propertyDefsDB, 'writePropertyDefinitions')
        jest.spyOn(propertyDefsDB, 'writeEventProperties')
    })

    afterEach(async () => {
        jest.restoreAllMocks()
        await closeHub(hub)
    })

    afterAll(() => {
        jest.useRealTimers()
    })

    const forSnapshot = <T>(value: T, overrides: Record<string, string> = {}): T => {
        return _forSnapshot(value, {
            overrides: {
                // Dates are tricky as it comes from the DB so is timezone dependent
                created_at: '<REPLACED_DATE>',
                last_seen_at: '<REPLACED_DATE>',
                ...overrides,
            },
        })
    }

    const sortPropertyDefinitions = (propertyDefinitions: PropertyDefinitionType[]) => {
        return propertyDefinitions.sort((a, b) =>
            `${a.project_id}-${a.name}-${a.type}`.localeCompare(`${b.project_id}-${b.name}-${b.type}`)
        )
    }

    const listSortedPropertyDefinitions = async (teamId: TeamId) => {
        // Helper to sort the responses to be deterministic
        return sortPropertyDefinitions(await propertyDefsDB.listPropertyDefinitions(teamId))
    }

    describe('property updates', () => {
        it('should write simple property defs to the DB', async () => {
            await ingester.handleKafkaBatch(
                createKafkaMessages([
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        properties: {
                            url: 'http://example.com',
                        },
                    }),
                ])
            )

            expect(propertyDefsDB.writeEventDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writePropertyDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writeEventProperties).toHaveBeenCalledTimes(1)

            expect(forSnapshot(await propertyDefsDB.listEventDefinitions(team.id))).toMatchSnapshot()

            expect(
                forSnapshot(await propertyDefsDB.listEventProperties(team.id), {
                    // NOTE: The event properties are not sorted by the DB so we need to sort them to be deterministic
                    id: '<REPLACED_NUMBER>',
                })
            ).toMatchSnapshot()

            expect(forSnapshot(await listSortedPropertyDefinitions(team.id))).toMatchSnapshot()
        })

        it('should only write the first seen property defs to the DB', async () => {
            await ingester.handleKafkaBatch(
                createKafkaMessages([
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        properties: {
                            url: 'http://example.com',
                        },
                    }),
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        properties: {
                            url: 2,
                        },
                    }),
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        properties: {
                            url: 5,
                        },
                    }),
                ])
            )

            expect(propertyDefsDB.writeEventDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writePropertyDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writeEventProperties).toHaveBeenCalledTimes(1)

            // Snapshot shows a String type as it was the first seen value
            expect(forSnapshot(await listSortedPropertyDefinitions(team.id))).toMatchSnapshot()
        })

        it('should batch multiple writes', async () => {
            await ingester.handleKafkaBatch(
                createKafkaMessages([
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        event: 'event1',
                        properties: {
                            url: 'http://example.com',
                        },
                    }),
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        event: 'event2',
                        properties: {
                            foo: 'bar',
                            test: 4,
                        },
                    }),
                ])
            )

            expect(propertyDefsDB.writeEventDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writePropertyDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writeEventProperties).toHaveBeenCalledTimes(1)

            // Snapshot shows a String type as it was the first seen value
            expect(forSnapshot(await listSortedPropertyDefinitions(team.id))).toMatchSnapshot()
        })

        it('should handle existing property defs', async () => {
            await ingester.handleKafkaBatch(
                createKafkaMessages([
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        event: 'event1',
                        properties: {
                            foo: 1,
                        },
                    }),
                ])
            )
            expect(forSnapshot(await listSortedPropertyDefinitions(team.id))).toMatchSnapshot()

            advanceTime(1000)

            await ingester.handleKafkaBatch(
                createKafkaMessages([
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        event: 'event1',
                        properties: {
                            other: 'property',
                        },
                    }),
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        event: 'event1',
                        properties: {
                            foo: 'changed!!',
                        },
                    }),
                ])
            )

            // Contains both properties but only one updated
            expect(forSnapshot(await listSortedPropertyDefinitions(team.id))).toMatchSnapshot()
        })
    })

    describe('group updates', () => {
        it('should create group properties from the $group_set property', async () => {
            await ingester.handleKafkaBatch(
                createKafkaMessages([
                    createClickHouseEvent({
                        team_id: team.id,
                        project_id: team.project_id,
                        event: '$groupidentify',
                        properties: {
                            $group_type: 'group-a',
                            $group_key: 'key-1',
                            $group_set: {
                                prop1: 'foo',
                                prop2: 2,
                            },
                        },
                    }),
                ])
            )

            expect(propertyDefsDB.writeEventDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writePropertyDefinitions).toHaveBeenCalledTimes(1)
            expect(propertyDefsDB.writeEventProperties).toHaveBeenCalledTimes(1)

            expect(forSnapshot(await propertyDefsDB.listEventDefinitions(team.id))).toMatchSnapshot()
            expect(forSnapshot(await listSortedPropertyDefinitions(team.id))).toMatchSnapshot()
        })
    })
})
