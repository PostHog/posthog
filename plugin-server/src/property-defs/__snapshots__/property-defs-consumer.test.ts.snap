// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyDefsConsumer group updates should create group properties from the $group_set property 1`] = `
[
  {
    "created_at": "<REPLACED_DATE>",
    "id": "<REPLACED-UUID-0>",
    "last_seen_at": "<REPLACED_DATE>",
    "name": "$groupidentify",
    "project_id": "2",
    "query_usage_30_day": null,
    "team_id": 2,
    "volume_30_day": null,
  },
]
`;

exports[`PropertyDefsConsumer group updates should create group properties from the $group_set property 2`] = `
[
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-0>",
    "is_numerical": false,
    "name": "$group_key",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-1>",
    "is_numerical": false,
    "name": "$group_type",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
  {
    "group_type_index": 0,
    "id": "<REPLACED-UUID-2>",
    "is_numerical": false,
    "name": "prop1",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 3,
    "volume_30_day": null,
  },
  {
    "group_type_index": 0,
    "id": "<REPLACED-UUID-3>",
    "is_numerical": true,
    "name": "prop2",
    "project_id": "2",
    "property_type": "Numeric",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 3,
    "volume_30_day": null,
  },
]
`;

exports[`PropertyDefsConsumer property updates should batch multipleÂ writes 1`] = `
[
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-0>",
    "is_numerical": false,
    "name": "foo",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-1>",
    "is_numerical": true,
    "name": "test",
    "project_id": "2",
    "property_type": "Numeric",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-2>",
    "is_numerical": false,
    "name": "url",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
]
`;

exports[`PropertyDefsConsumer property updates should handle existing property defs 1`] = `
[
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-0>",
    "is_numerical": true,
    "name": "foo",
    "project_id": "2",
    "property_type": "Numeric",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
]
`;

exports[`PropertyDefsConsumer property updates should handle existing property defs 2`] = `
[
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-0>",
    "is_numerical": true,
    "name": "foo",
    "project_id": "2",
    "property_type": "Numeric",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-1>",
    "is_numerical": false,
    "name": "other",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
]
`;

exports[`PropertyDefsConsumer property updates should only write the first seen property defs to the DB 1`] = `
[
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-0>",
    "is_numerical": false,
    "name": "url",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
]
`;

exports[`PropertyDefsConsumer property updates should write simple property defs to the DB 1`] = `
[
  {
    "created_at": "<REPLACED_DATE>",
    "id": "<REPLACED-UUID-0>",
    "last_seen_at": "<REPLACED_DATE>",
    "name": "$pageview",
    "project_id": "2",
    "query_usage_30_day": null,
    "team_id": 2,
    "volume_30_day": null,
  },
]
`;

exports[`PropertyDefsConsumer property updates should write simple property defs to the DB 2`] = `
[
  {
    "event": "$pageview",
    "id": "<REPLACED_NUMBER>",
    "project_id": "2",
    "property": "url",
    "team_id": 2,
  },
]
`;

exports[`PropertyDefsConsumer property updates should write simple property defs to the DB 3`] = `
[
  {
    "group_type_index": null,
    "id": "<REPLACED-UUID-0>",
    "is_numerical": false,
    "name": "url",
    "project_id": "2",
    "property_type": "String",
    "property_type_format": null,
    "query_usage_30_day": null,
    "team_id": 2,
    "type": 1,
    "volume_30_day": null,
  },
]
`;
